client/state/states-list/reducer.js

```diff
@@ -0,0 +1,32 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+
+/**
+ * Internal dependencies
+ */
+import {
+	STATES_LIST_RECEIVE,
+	STATES_LIST_REQUEST,
+	STATES_LIST_REQUEST_FAILURE,
+} from 'state/action-types';
+import { statesListSchema } from './schema';
+import { createReducer } from 'state/utils';
+
+// Stores the complete list of states, indexed by locale key
+export const statesList = createReducer( {}, {
+	[ STATES_LIST_RECEIVE ]: ( state, action ) => ( { ...state, [ action.countryCode ]: action.statesList } ),
```

Comment:
It'd depend on the UI requirements, but do we ever have a need for finding a state name by it's country code, state code pair? If so, how would we implement that as a selector given this state shape? Would probably require iterating over all states 'til we find the one we're looking for. For these cases, [Lodash's `_.keyBy`](https://lodash.com/docs/4.16.2#keyBy) can be a nice way to shape the state, keying it by the code.

```js
[ STATES_LIST_RECEIVE ]: ( state, action ) => ( { ...state, [ action.countryCode ]: keyBy( action.statesList, 'code' ) } )
```

```js
export function getStatesListState( state, countryCode, stateCode ) {
	return get( state.statesList.statesList, [ countryCode, stateCode ], null );
}
```

_(See also: [`_.get`](https://lodash.com/docs/4.16.2#get))_

Again, this depends on the requirements. If we're only using it to generate `<select>` dropdowns, storing the array as you've done here may in fact be the best solution.